### PR TITLE
Chan x/y placement cost factors using prefix sum

### PR DESCRIPTION
**Pull request by soheilshahrouz**
Implements the prefix sum idea used for chanz in #2781 for chanx/y.

### CODE REVIEW

File: vpr/src/place/net_cost_handler.cpp [hunk 1]
```diff
@@ -153,81 +153,36 @@ NetCostHandler::NetCostHandler(const t_placer_opts& placer_opts,
 }
 
 void NetCostHandler::alloc_and_load_chan_w_factors_for_place_cost_(float place_cost_exp) {
-    auto& device_ctx = g_vpr_ctx.device();
+    const auto& device_ctx = g_vpr_ctx.device();
 
     const int grid_height = device_ctx.grid.height();
     const int grid_width = device_ctx.grid.width();
 
     /* Access arrays below as chan?_place_cost_fac_(subhigh, sublow). Since subhigh must be greater than or
```
Comment (vaughnbetz) [inline vpr/src/place/net_cost_handler.cpp:161]: After checking QoR (if bad, we won't merge), update these comments.

File: vpr/src/place/net_cost_handler.cpp [hunk 2]
```diff
@@ -153,81 +153,36 @@ NetCostHandler::NetCostHandler(const t_placer_opts& placer_opts,
...
+    acc_chany_width_ = vtr::NdOffsetMatrix<int, 1>({{{-2, grid_width}}});
 
     // First compute the number of tracks between channel high and channel low, inclusive.
-    chanx_place_cost_fac_[-1][-1] = 0;
-
-    for (int high = 0; high < grid_height; high++) {
-        chanx_place_cost_fac_[high][high] = (float)device_ctx.chan_width.x_list[high];
-        for (int low = -1; low < high; low++) {
-            chanx_place_cost_fac_[high][low] = chanx_place_cost_fac_[high - 1][low] + (float)device_ctx.chan_width.x_list[high];
+    acc_chanx_width_[-2] = 0;
```
Comment (vaughnbetz) [inline vpr/src/place/net_cost_handler.cpp:168]: I think we can get rid of the [-2] entry, and the = -1 should be = 0.

File: vpr/src/place/net_cost_handler.cpp [hunk 3]
```diff
@@ -153,81 +153,36 @@ NetCostHandler::NetCostHandler(const t_placer_opts& placer_opts,
...
-             * factor, and extremely bad placer behaviour. Hence we change *
-             * this to a small (1 track) channel capacity instead.         */
-            if (chanx_place_cost_fac_[high][low] == 0.0f) {
-                VTR_LOG_WARN("CHANX place cost fac is 0 at %d %d\n", high, low);
-                chanx_place_cost_fac_[high][low] = 1.0f;
-            }
-
-            chanx_place_cost_fac_[high][low] = (high - low + 1.) / chanx_place_cost_fac_[high][low];
-            chanx_place_cost_fac_[high][low] = pow((double)chanx_place_cost_fac_[high][low], (double)place_cost_exp);
+    acc_chany_width_[-2] = 0;
```
Comment (vaughnbetz) [inline vpr/src/place/net_cost_handler.cpp:177]: Same comment.

File: vpr/src/place/net_cost_handler.cpp [hunk 4]
```diff
@@ -1520,11 +1477,10 @@ double NetCostHandler::get_net_per_layer_bb_cost_(ClusterNetId net_id , bool use
...
          */
 
-        ncost += (bb[layer_num].xmax - bb[layer_num].xmin + 1) * crossing
-                 * chanx_place_cost_fac_[bb[layer_num].ymax][bb[layer_num].ymin - 1];
+        const auto[chanx_cost_fac, chany_cost_fac] = get_chan_place_fac_(bb[layer_num]);
+        ncost += (bb[layer_num].xmax - bb[layer_num].xmin + 1) * crossing * chanx_cost_fac;
 
-        ncost += (bb[layer_num].ymax - bb[layer_num].ymin + 1) * crossing
-                 * chany_place_cost_fac_[bb[layer_num].xmax][bb[layer_num].xmin - 1];
+        ncost += (bb[layer_num].ymax - bb[layer_num].ymin + 1) * crossing * chany_cost_fac;
```
Comment (vaughnbetz) [inline vpr/src/place/net_cost_handler.cpp:1483]: Soheil to Soheil comment: factor out crossing (good idea!) to remove a floating point op.  I just learned floating point ops can't be reordered, as the compiler guarantees the round off.

File: vpr/src/place/net_cost_handler.h [hunk 5]
```diff
@@ -511,4 +512,17 @@ class NetCostHandler {
      */
     double get_net_wirelength_from_layer_bb_(ClusterNetId net_id);
 
+    template<typename BBT>
+    std::pair<double, double> get_chan_place_fac_(const BBT& bb) {
```
Comment (vaughnbetz) [inline vpr/src/place/net_cost_handler.h:516]: If this has good QoR, comment this routine to explain how this works.  Also comment the template parameter (what are the types this handles).

File: vpr/src/place/net_cost_handler.h [hunk 6]
```diff
@@ -511,4 +512,17 @@ class NetCostHandler {
...
+        const int total_chanx_width = acc_chanx_width_[bb.ymax] - acc_chanx_width_[bb.ymin - 2];
+        const double inverse_average_chanx_width = (bb.ymax - bb.ymin + 2.0) / total_chanx_width;
+        const double inverse_average_chanx_width_sharpened = std::pow(inverse_average_chanx_width, (double)placer_opts_.place_cost_exp);
```
Comment (vaughnbetz) [inline vpr/src/place/net_cost_handler.h:519]: You could try experimenting with removing the pow and see if it yields any cpu speedup; we run with a place_cost_exp of 1 pretty much all the time so if the pow is costing us we could remove it.

File: vpr/src/place/net_cost_handler.h [hunk 7]
```diff
@@ -511,4 +512,17 @@ class NetCostHandler {
      */
     double get_net_wirelength_from_layer_bb_(ClusterNetId net_id);
 
+    template<typename BBT>
+    std::pair<double, double> get_chan_place_fac_(const BBT& bb) {
+        const int total_chanx_width = acc_chanx_width_[bb.ymax] - acc_chanx_width_[bb.ymin - 2];
```
Comment (vaughnbetz) [inline vpr/src/place/net_cost_handler.h:517]: Could get rid of the -2 if we don't make the bounding box include channels on the left/below edges (just have to comment it).

**Review by vaughnbetz — COMMENTED** [review]
Looks good; if the QoR is OK we can merge, but I suggest some tweaks.

**Comment by soheilshahrouz** [comment]
titan_quick_qor over 3 seeds

| Metric                                              | master.txt | feature.txt   |
|-----------------------------------------------------|------------|---------------|
| vtr_flow_elapsed_time                               | 1          | 1.001301668   |
| num_LAB                                             | 1          | 1             |
| num_DSP                                             | 1          | 1             |
| num_M9K                                             | 1          | 1             |
| num_M144K                                           | 1          | 1             |
| max_vpr_mem                                         | 1          | 0.9999549581  |
| num_pre_packed_blocks                               | 1          | 1             |
| num_post_packed_blocks                              | 1          | 1             |
| device_grid_tiles                                   | 1          | 1             |
| pack_time                                           | 1          | 1.007941195   |
| placed_wirelength_est                               | 1          | 0.9979723975  |
| place_time                                          | 1          | 1.001301333   |
| placed_CPD_est                                      | 1          | 1.019405141   |
| routed_wirelength                                   | 1          | 0.9981684731  |
| critical_path_delay                                 | 1          | 1.020593524   |
| geomean_nonvirtual_intradomain_critical_path_delay  | 1          | 0.9936476626  |
| crit_path_route_time                                | 1          | 1.016755021   |

**Comment by soheilshahrouz** [comment]
> QoR seems OK except cpd is up 2%. Probably noise but what does the circuit by circuit result look like?

Here [the link](https://docs.google.com/spreadsheets/d/13zDcBb2PW_hKbs-b3iml_vAqMmYfRwXWgxeKeAhI7iM/edit?usp=sharing) to circuit by circuit comparison. In the most extreme case, it is increased by ~40%.

**Comment by vaughnb-cerebras** [comment]
Thanks.  It looks like there is just a couple of circuits with noise. I'm OK with merging.

**Comment by vaughnb-cerebras** [comment]
Suggest making the changes listed in the code review, doing a quick re-check to be safe on QoR, and then merging.

**Comment by soheilshahrouz** [comment]
titan_quick_qor on 464dd62

| Metric                                           | Baseline | Value        |
|--------------------------------------------------|----------|--------------|
| vtr_flow_elapsed_time                            | 1        | 1.032769487  |
| num_LAB                                          | 1        | 1            |
| num_DSP                                          | 1        | 1            |
| num_M9K                                          | 1        | 1            |
| num_M144K                                        | 1        | 1            |
| max_vpr_mem                                      | 1        | 0.9999541852 |
| num_pre_packed_blocks                            | 1        | 1            |
| num_post_packed_blocks                           | 1        | 1            |
| device_grid_tiles                                | 1        | 1            |
| pack_time                                        | 1        | 1.011271438  |
| placed_wirelength_est                            | 1        | 0.997215378  |
| place_time                                       | 1        | 1.043569677  |
| placed_CPD_est                                   | 1        | 1.009797219  |
| routed_wirelength                                | 1        | 0.997584742  |
| critical_path_delay                              | 1        | 1.007112939  |
| geomean_nonvirtual_intradomain_critical_path_delay | 1      | 0.995636464  |
| crit_path_route_time                             | 1        | 1.011247505  |


The increase in CPD is now 1%, but palcement time increased by 4%.

**Comment by vaughnbetz** [comment]
Probably really 3% as pack and route time are up 1%. Any ideas to optimize? Take out the pow?

Special case all channels equal width (check and store) and compute the pow and reciprocal in advance?

**Comment by soheilshahrouz** [comment]
titan_quick_qor
Comparison between master (045a9e81) and this branch (49d43968)

It seems that multiplying `crossing` only once at the end has reduced the placement time.

| Metric                                     | master.txt | feature.txt     |
|--------------------------------------------|------------|-----------------|
| vtr_flow_elapsed_time                      | 1          | 0.9960061822    |
| num_LAB                                    | 1          | 1               |
| num_DSP                                    | 1          | 1               |
| num_M9K                                    | 1          | 1               |
| num_M144K                                  | 1          | 1               |
| max_vpr_mem                                | 1          | 1.000009388     |
| num_pre_packed_blocks                      | 1          | 1               |
| num_post_packed_blocks                     | 1          | 1               |
| device_grid_tiles                          | 1          | 1               |
| pack_time                                  | 1          | 0.9980662387    |
| placed_wirelength_est                      | 1          | 1.001936399     |
| place_time                                 | 1          | 0.9943848951    |
| placed_CPD_est                             | 1          | 0.9966367297    |
| routed_wirelength                          | 1          | 1.002183219     |
| critical_path_delay                        | 1          | 0.993269001     |
| geomean_nonvirtual_intradomain_CPD         | 1          | 0.9936908393    |
| crit_path_route_time                       | 1          | 0.9978216647    |

**Comment by soheilshahrouz** [comment]
For reciprocal calculation, we can use [fast inverse sequare root](https://en.wikipedia.org/wiki/Fast_inverse_square_root) algorithm and multiply the result by itself to get the reciprocal.

**Comment by vaughnbetz** [comment]
Cool, thanks. I think we should merge this; the fast reciprocal square root is probably not worth the effort/complexity.

There are some QoR failures, but they all seem to be on tiny circuits so I think OK, and the one in basic is a big win.

**Comment by soheilshahrouz** [comment]
@vaughnbetz ready to merge into master

**Pull request merged**